src/main/scala/firrtl/passes/NewLowerTypes.scala

```diff
@@ -0,0 +1,439 @@
+
+        val newPortType = newMemReference.tpe.asInstanceOf[BundleType].fields.find(_.name == port.name).get.tpe
+        val newPortAccess = SubField(newMemReference, port.name, newPortType)
+
+        port.tpe.asInstanceOf[BundleType].fields.map { portField =>
+          val isDataField = portField.name == "data" || portField.name == "wdata" || portField.name == "rdata"
+          val isMaskField = portField.name == "mask" || portField.name == "wmask"
+          val isDataOrMaskField = isDataField || isMaskField
+          val oldFieldRefs = if(memWasRenamed && isDataOrMaskField) {
+            // there might have been multiple different fields which now alias to the same lowered field.
+            val oldPortFieldBaseRef = oldPortRef.field(portField.name)
+            refSuffixes.map(s => oldPortFieldBaseRef.copy(component = oldPortFieldBaseRef.component ++ s))
+          } else {
+            List(oldPortRef.field(portField.name))
+          }
+
+          val newPortType = if(isDataField) { newMem.dataType } else { portField.tpe }
+          val newPortFieldAccess = SubField(newPortAccess, portField.name, newPortType)
+
+          // record renames only for the data field which is the only port field of non-ground type
+          val newPortFieldRef = newPortRef.field(portField.name)
+          if(memWasRenamed && isDataOrMaskField) {
+            oldFieldRefs.foreach { o => renameMap.record(o, newPortFieldRef) }
+          }
+
+          val oldFieldStringRef = extractGroundTypeRefString(oldFieldRefs)
+          (oldFieldStringRef, newPortFieldAccess)
+        }
+      }
+      (newMem, subFields)
+    }
+
+    (newMemAndSubFields.map(_._1), newMemAndSubFields.flatMap(_._2))
+  }
+
+  private def memBundle(mem: DefMemory): Field = mem.dataType match {
+    case _: GroundType => Field(mem.name, Default, mem.dataType)
+    case _: BundleType | _: VectorType =>
+      val subMems = getFields(mem.dataType).map(f => mem.copy(name = f.name, dataType = f.tpe))
+      val fields = subMems.map(memBundle)
+      Field(mem.name, Default, BundleType(fields))
+  }
+
+  private def recordRenames(fieldToRefs: Seq[(Field, Seq[ReferenceTarget])], renameMap: RenameMap, parent: ParentRef):
+  Unit = {
+    // TODO: if we group by ReferenceTarget, we could reduce the number of calls to `record`. Is it worth it?
+    fieldToRefs.foreach { case(field, refs) =>
+      val fieldRef = parent.ref(field.name)
+      refs.foreach{ r => renameMap.record(r, fieldRef) }
+    }
+  }
+
+  private def extractGroundTypeRefString(refs: Seq[ReferenceTarget]): String =
+  if(refs.isEmpty) { "" } else {
+    // Since we depend on ExpandConnects any reference we encounter will be of ground type
+    // and thus the one with the longest access path.
+    refs.reduceLeft((x,y) => if (x.component.length > y.component.length) x else y)
+    // convert references to strings relative to the module
+      .serialize.dropWhile(_ != '>').tail
+  }
```

Comment:
```suggestion
    if(refs.isEmpty) { "" } else {
      // Since we depend on ExpandConnects any reference we encounter will be of ground type
      // and thus the one with the longest access path.
      refs.reduceLeft((x,y) => if (x.component.length > y.component.length) x else y)
      // convert references to strings relative to the module
        .serialize.dropWhile(_ != '>').tail
    }
```